### ISSUE TITLE
Committed bloom filters for improved wallet performance and SPV security

### ISSUE THREAD

**Issue by Giszmo** [bitcoin-wallet/bitcoin-wallet#358]
I always feel sorry when I have to bring up the poor privacy aspects of this wonderful wallet but with the current state of Bloom filters being sent to several full nodes, these full nodes can know almost exactly which addresses constitute accounts.

[Committed bloom filters for improved wallet performance and SPV security](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2016-May/012636.html) did in that respect not get anywhere near the attention it would deserve.

By my estimate [and experiments](https://github.com/Giszmo/TransactionFinder) it would cost "only" 3GB on the user's phone to have an index of all **relevant** transactions. What is relevant depends on the wallet but for the Schildbach wallet only P2PKH transactions would be relevant, right? So instead of waiting hours to resync the blockchain, users would wait minutes to parse the local index when an address of unknown age gets added and only relevant blocks could get downloaded in full.

**Comment by schildbach** [comment]
Thanks for reminding us of that proposal. It would be best to post it to the bitcoinj mailing list, because the BIP would need to implemented in bitcoinj. Bitcoinj also supports and actively uses P2PK and supports but not actively uses P2SH.

**Comment by Giszmo** [comment]
[done](https://github.com/bitcoinj/bitcoinj/issues/1313). Still I believe it has to be documented that this wallet is ~~leaking~~ shouting to the world the user's wallet. This is a fundamental issue of this wallet and if bitcoinJ doesn't fix it, the wallet has to fix it. (Yeah, I know, same team but this is not some minor glitch.)

Bitcoinj's use of P2PK was rightfully criticized by @ManfredKarrer and yes, needing 3GB might be manageable while needing 3GB per scope (P2PKH, P2PK, P2SH and P2WPKH) can quickly switch from being a great improvement over running a full node to not worth the hassle, although we are in a sweet spot where phones would support 32GB of microSD cards but can't handle the full chain. Maybe next year this will be a moot point.

**Comment by ManfredKarrer** [comment]
@schildbach What is the usage of P2PK transaction? I know it is supported by BitcoinJ but I am not aware of anyone still using that.
To simply not add the P2PK data to the filter will already remove the most critical flaw. 
To persist the nonce and reuse that at every startup is another super simple fix. I am aware that there are still more issues open but those 2 steps are at least so cheap that I cannot see a reason why not to apply them.

**Comment by Giszmo** [comment]
> @schildbach What is the usage of P2PK transaction? I know it is supported by BitcoinJ but I am not aware of anyone still using that.
> To simply not add the P2PK data to the filter will already remove the most critical flaw. 
> To persist the nonce and reuse that at every startup is another super simple fix. I am aware that there are still more issues open but those 2 steps are at least so cheap that I cannot see a reason why not to apply them.

The #1 reason of not applying these is that it leaves the user in falsely perceived security. As long as you are interested in a set of addresses *that are connected by joins and chaining* an adversary can almost trivially figure out which addresses you are interested in. We should stop considering Bloom filters as a privacy feature and only treat it as a clever compression of an address list.

**Comment by ManfredKarrer** [comment]
I agree but current situation is that nearly nobody beside a few devs are aware that the privacy of BitcoinJ powered wallets is 100% broken. To get the bloomfilters fixed would be great but it does not seem realistic that we get that soon. So what we can do now? At least increasing them a bit seems to me better than doing nothing and just wait.

**Comment by Giszmo** [comment]
@ManfredKarrer you are posting in a kind of unrelated issue. This issue is about a Bloom filter based blockchain index.

The current Bloom filter use is broken beyond repair when it comes to privacy. Lets make it obsolete with a transactions index.

**Comment by schildbach** [comment]
@ManfredKarrer Wallet rotations were done via P2PK, though I think it was changed to P2PKH a while ago. Still, I guess a lot of wallet created before Aug 2013 contain such transactions (e.g. mine). I'm not sure wether rotation transactions would fit the filter anyway, provided they should match also because they are sending not only to be also *from* the wallet.

Plus, we'll likely need to support segwit which comes with more forms of scripts, doesn't it?

Your two suggestions sound like no-brainers. Are there any open PRs I overlooked? @Giszmo I wouldn't care about "falsely perceived privacy". The app doesn't claim to be super private, although I wish it could.

Also I think P2P link encryption/authentication would be good to have, because users with privacy needs could then run their own full node to connect to. (They already can, but the connection can be eavesdropped.)

**Comment by ManfredKarrer** [comment]
@schildbach I have implemented those changes in my branch (https://github.com/bitsquare/bitcoinj) but as I need serialisation support I am stuck to an older version (planned in future to get back to master, but need to change java serialisation for protobuffer first...). So no I did not do a PR...

**Comment by 3824409A28** [comment]
Withdraw funds to my bank in vietnam.iam going to do first?

**Comment by schildbach** [comment]
Sorry for resurrecting this topic. While doing the implementation for P2WPKH I though a lot about bloom filters. I discovered we cannot remove pubkeys from our bloom filters because not only would we then not match incoming P2PK transactions, but also not *outgoing* transactions that don't send change back into our wallet. Most prominently, this affects empty-wallet transactions while replaying. In other words, a majority of the current wallets couldn't be replayed without missing transactions.

The other thing is, there is still no alternative for bloom filters on the horizon. BIP157/158 are a nice concept, but:

- Doesn't work for pending transactions. This degrades them from enabled-by-default to a "more private, but also much slower" option which I believe will only be used by a handful of people.
- Filters still not committed to blockchain. And who knows if they ever will? And if they are, we can only only offer the option to wallets created after that date.
- Most important: They don't scale with wallets, unless we commit multiple versions with different false positive rates. This would force us to reuse addresses at some point, so basically we're back to the same level of privacy as bloom filters.

What I like about BIP157/158:

- The network cannot lie about omission of transactions.
- Rather than Pubkey Hashes, entire scriptPubKeys are put into the filter. This makes them kind of future proof.

**Comment by ManfredKarrer** [comment]
When I tried to remove pubkeys from filters I encountered problems as well. Have not investigeted further what was the exact reason but it was a P2SH multisig without change. So either the reason was as you mentioned he case when you dont get a change or that Multisig txs also require pubkeys in the bloom filter.

From my persepective the only feasible way to live with BIP 37 is to only connect to trusted nodes. Best your own node or if the use does not have one nodes operated by trusted peers. Of course not a great solution either but seems the best of all the imperfect ways how to deal with light wallets.

**Comment by Giszmo** [comment]
BIP158 is being used in Wasabi and other projects and although there is no commitment by the miners, it should be possible to get more than one source of filters. The privacy implications of sending the complete list of addresses (bloomfiltered or not) to random full nodes is just disastrous (or did I miss some development here?)

I know users can configure their own full nodes but 99.99% of users don't and some might but not with ssl, ...

I'm asking now because I'm working on a [new feature](https://gitlab.com/walletscrutiny/walletScrutinyCom/-/issues/99) for my project walletscrutiny. I want to show warnings about reviewed wallets.

(Most importantly warnings that call in question the given verdict. Wallets that have custodial and non-custodial accounts for example will be listed as non-custodial with a warning about custodial parts)

But I want to use the feature to warn about other issues, too, and in the case of "Bitcoin Wallet", privacy is a big concern. It would be a "medium severity" warning.

Is my assessment right that in all likelihood almost all addresses of almost all users of this wallet are known to blockchain analysis companies assuming those run half of all reachable full nodes?

**Comment by schildbach** [comment]
For BIP37 users they could group addresses I guess. In practise, II doubt that it will be done so widely – matching a filter on the entire blockchain takes quite some effort, and there is plenty of filter candidates.

I think the traffic saving from BIP158 when used with at least 3 sources of filters is negliable compared to always receiving full blocks. But I'm willing to try out if client-side filtering gets implemented in bitcoinj.

Re-reading this issue I thought for completeness I should mention that I think we could improve the privacy of server-side filtering a lot if we added only one item per address to the filter. That would be the scriptPubKey of the address, using [the same rules as in BIP158](https://github.com/bitcoin/bips/blob/master/bip-0158.mediawiki#contents). That would free us from the requirement to also add the pubKey per address, and at the same time make the filter rules future-proof. The change would require a new BIP and support in Bitcoin Core however.

**Comment by Giszmo** [comment]
> For BIP37 users they could cluster addresses I guess. In practise, II doubt that it will be done so widely – matching a filter on the entire blockchain takes quite some effort, and there is plenty of filter candidates.

So filters generated by clients to monitor their wallets can be stored trivially for later analysis by those Chain Analysis companies. Clustering addresses and querying the same cluster from the same node would help a bit but it's still a privacy disaster.

> I think the traffic saving from BIP158 when used with at least 3 sources of filters is negliable compared to always receiving full blocks. But I'm willing to try out if client-side filtering gets implemented in bitcoinj.

If those three sources confirm each other's filter hash, it should not be necessary to download all three. Only when there is a discrepancy. And even if you have to download all three, the savings from neutrino are way more than 66%.

Dealing with neutrino block filters client side would also enable the wallet to add key import / cold wallet spending / hardware wallet integration without prohibitively long scanning of the whole blockchain.
